agent/consul/leader_connect_ca.go

```diff
@@ -0,0 +1,1082 @@
+	"github.com/hashicorp/go-hclog"
+	uuid "github.com/hashicorp/go-uuid"
+)
+
+type CAState string
+
+const (
+	CAStateUninitialized     CAState = "UNINITIALIZED"
+	CAStateInitializing              = "INITIALIZING"
+	CAStateReady                     = "READY"
+	CAStateRenewIntermediate         = "RENEWING"
+	CAStateReconfig                  = "RECONFIGURING"
+)
+
+// CAManager is a wrapper around CA operations such as updating roots, an intermediate
+// or the configuration. All operations should go through the CAManager in order to
+// avoid data races.
+type CAManager struct {
+	srv    *Server
+	logger hclog.Logger
+
+	providerLock sync.RWMutex
+	// provider is the current CA provider in use for Connect. This is
+	// only non-nil when we are the leader.
+	provider ca.Provider
+	// providerRoot is the CARoot that was stored along with the ca.Provider
+	// active. It's only updated in lock-step with the provider. This prevents
+	// races between state updates to active roots and the fetch of the provider
+	// instance.
+	providerRoot *structs.CARoot
+
+	// stateLock protects the internal state used for administrative CA tasks.
+	stateLock         sync.Mutex
+	state             CAState
+	primaryRoots      structs.IndexedCARoots // The most recently seen state of the root CAs from the primary datacenter.
+	actingSecondaryCA bool                   // True if this datacenter has been initialized as a secondary CA.
+}
+
+func NewCAManager(srv *Server) *CAManager {
+	return &CAManager{
+		srv:    srv,
+		logger: srv.loggers.Named(logging.Connect),
+		state:  CAStateUninitialized,
+	}
+}
+
+// setState attempts to update the CA state to the given state.
+// If the current state is not READY, this will fail. The only exception is when
+// the current state is UNINITIALIZED, and the function is called with CAStateInitializing.
+func (c *CAManager) setState(newState CAState, validateState bool) error {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+	state := c.state
+
+	if !validateState || state == CAStateReady || (state == CAStateUninitialized && newState == CAStateInitializing) {
+		c.state = newState
+	} else {
+		return fmt.Errorf("CA is already in state %q", state)
+	}
+	return nil
+}
+
+// setPrimaryRoots updates the most recently seen roots from the primary.
+func (c *CAManager) setPrimaryRoots(newRoots structs.IndexedCARoots) error {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+
+	if c.state == CAStateInitializing || c.state == CAStateReconfig {
+		c.primaryRoots = newRoots
+	} else {
+		return fmt.Errorf("Cannot update primary roots in state %q", c.state)
+	}
+	return nil
+}
+
+func (c *CAManager) getPrimaryRoots() structs.IndexedCARoots {
+	c.stateLock.Lock()
+	defer c.stateLock.Unlock()
+	return c.primaryRoots
+}
+
+// initializeCAConfig is used to initialize the CA config if necessary
+// when setting up the CA during establishLeadership
+func (c *CAManager) initializeCAConfig() (*structs.CAConfiguration, error) {
```

Comment:
maybe add a comment about how you should hold the state "meta lock" before calling this.